cleanrl/ppo.py

```diff
@@ -159,15 +166,24 @@ def get_action_and_value(self, x, action=None):
     # env setup
     envs = gym.vector.SyncVectorEnv(
         [make_env(args.env_id, args.seed + i, i, args.capture_video, run_name) for i in range(args.num_envs)]
-    )
+    )  # type:ignore[abstract]
     assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+    # Handling gym shapes being Optionals (variant 1)
+    # Personally i'd prefer the asserts
+    assert isinstance(envs.single_observation_space.shape, tuple), "shape of observation space must be defined"
+    assert isinstance(envs.single_action_space.shape, tuple), "shape of action space must be defined"
+
+    # Handling gym shapes being Optionals (variant 2)
+    # Once could also cast inside each call but in my eyes that's not conducive to readability
+    obs_space_shape = cast(tuple[int, ...], envs.single_observation_space.shape)
+    action_space_shape = cast(tuple[int, ...], envs.single_action_space.shape)
```

Comment:
Gym spaces can in theory return `None` shapes. Mypy will complain about this when concatenating the shape tuples later.
- Option 1 is to use a cast either once here or every time the spaces are accessed. I don't think that's very readable.
- Option 2 is to assert that the space shapes are tuples. Doing it once here fixes all errors for the rest of the code. Since there's an assert in this place already anyway I think this is the better option.

Reply:
Adding `from __future__ import annotation` at the top would guarantee the 3.7+ compatibility, and make dropping support later on quite easy.